In [ ]:
import pandas as pd


In [ ]:
df = pd.read_csv("/content/tips.csv") # read csv file
df

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [ ]:
df.head() # displays first 5 rows of the data

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [ ]:
df.tail() # displays last 5 rows of the data

total_bill   tip     sex smoker   day    time  size
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

In [ ]:
# another way
df.head(3) # first 3 rows

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3

In [ ]:
df.sample(3) # generates 3 random sample from the table

total_bill  tip     sex smoker  day    time  size
29        19.65  3.0  Female     No  Sat  Dinner     2
73        25.28  5.0  Female    Yes  Sat  Dinner     2
227       20.45  3.0    Male     No  Sat  Dinner     4

In [ ]:
# eveery time you run df.sample you will get new samples. If we want to get one fixed sample we need seed in pandas sedd means random_state
# we should set random_satate >=0

In [ ]:
df.sample(3, random_state=2023) # preview herna ko lagi

total_bill   tip     sex smoker  day    time  size
154       19.77  2.00    Male     No  Sun  Dinner     4
4         24.59  3.61  Female     No  Sun  Dinner     4
30         9.55  1.45    Male     No  Sat  Dinner     2

# Data Profiling and Inspection

In [ ]:
df.info() # displays information about the table

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    object 
 3   smoker      244 non-null    object 
 4   day         244 non-null    object 
 5   time        244 non-null    object 
 6   size        244 non-null    int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 13.5+ KB


In [ ]:
type(df) # df vanne object ko datatype

pandas.core.frame.DataFrame

In [ ]:
df.shape #(no.of rows, no.of columns)

(244, 7)

In [ ]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [ ]:
# what are the columns name
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size'], dtype='object')

In [ ]:
df.dtypes # data types of each column

total_bill    float64
tip           float64
sex            object
smoker         object
day            object
time           object
size            int64
dtype: object

In [ ]:
# data contains 244 rows and if df.info() shows 244 not null in each column so we have missing data.

In [ ]:
# How to verify?
df.isna().sum()

total_bill    0
tip           0
sex           0
smoker        0
day           0
time          0
size          0
dtype: int64

In [ ]:
# descriptive statistics of the data
df.describe()

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

# For total_bill column:
* count = 244 : There are 244 not null rows in total_bill column
* mean = 19.78 : The average total_bill is 19.78
* the standard deviation of total bill is 8.90
* The minimum total bill is 3.07
* 25% of the data points lies below 13.34 for total bill column
* 50% of the points lies below 17.79 for total bill column
* 75% of the data points lies below 24.12 for total bill column
* The maximum total bill occured is 50.81

In [ ]:
#  I want to describe object type too.
df.describe(include='object')

sex smoker  day    time
count    244    244  244     244
unique     2      2    4       2
top     Male     No  Sat  Dinner
freq     157    151   87     176

* There are 244 not null rows in sex column
* the sex colun contains two unique values
* The most occuring (top) value is male with 157 times/frequency

In [ ]:
len(df['sex'])

244

In [ ]:
df['sex'].unique()

array(['Female', 'Male'], dtype=object)

In [ ]:
df['sex'].value_counts()

Male      157
Female     87
Name: sex, dtype: int64

In [ ]:
%%time
df['sex']

CPU times: user 61 µs, sys: 5 µs, total: 66 µs
Wall time: 71.5 µs


0      Female
1        Male
2        Male
3        Male
4      Female
        ...  
239      Male
240    Female
241      Male
242      Male
243    Female
Name: sex, Length: 244, dtype: object

In [ ]:
len(df['day'])

244

In [ ]:
df['day'].unique()

array(['Sun', 'Sat', 'Thur', 'Fri'], dtype=object)

In [ ]:
df['day'].value_counts()

Sat     87
Sun     76
Thur    62
Fri     19
Name: day, dtype: int64

# **loc**

In [ ]:
# on average who tips more smoker or non-smoker?
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [ ]:
df['smoker']=='Yes'

0      False
1      False
2      False
3      False
4      False
       ...  
239    False
240     True
241     True
242    False
243    False
Name: smoker, Length: 244, dtype: bool

In [ ]:
df['smoker']=='No'

0       True
1       True
2       True
3       True
4       True
       ...  
239     True
240    False
241    False
242     True
243     True
Name: smoker, Length: 244, dtype: bool

In [ ]:
df[df['smoker']=='Yes']

total_bill   tip     sex smoker  day    time  size
56        38.01  3.00    Male    Yes  Sat  Dinner     4
58        11.24  1.76    Male    Yes  Sat  Dinner     2
60        20.29  3.21    Male    Yes  Sat  Dinner     2
61        13.81  2.00    Male    Yes  Sat  Dinner     2
62        11.02  1.98    Male    Yes  Sat  Dinner     2
..          ...   ...     ...    ...  ...     ...   ...
234       15.53  3.00    Male    Yes  Sat  Dinner     2
236       12.60  1.00    Male    Yes  Sat  Dinner     2
237       32.83  1.17    Male    Yes  Sat  Dinner     2
240       27.18  2.00  Female    Yes  Sat  Dinner     2
241       22.67  2.00    Male    Yes  Sat  Dinner     2

[93 rows x 7 columns]

In [ ]:
df.loc[df['smoker']=='Yes']

total_bill   tip     sex smoker  day    time  size
56        38.01  3.00    Male    Yes  Sat  Dinner     4
58        11.24  1.76    Male    Yes  Sat  Dinner     2
60        20.29  3.21    Male    Yes  Sat  Dinner     2
61        13.81  2.00    Male    Yes  Sat  Dinner     2
62        11.02  1.98    Male    Yes  Sat  Dinner     2
..          ...   ...     ...    ...  ...     ...   ...
234       15.53  3.00    Male    Yes  Sat  Dinner     2
236       12.60  1.00    Male    Yes  Sat  Dinner     2
237       32.83  1.17    Male    Yes  Sat  Dinner     2
240       27.18  2.00  Female    Yes  Sat  Dinner     2
241       22.67  2.00    Male    Yes  Sat  Dinner     2

[93 rows x 7 columns]

In [ ]:
# calculate mean of tip
df.loc[df['smoker']=='Yes'].tip.mean()

3.008709677419355

In [ ]:
df.loc[df['smoker']=='No'].tip.mean()

2.9918543046357615

In [ ]:
#  on average in which day you will get more tip?


In [ ]:
df['day'].unique()

array(['Sun', 'Sat', 'Thur', 'Fri'], dtype=object)

In [ ]:
df.loc[df['day']=='Sun'].tip.mean()

3.2551315789473687

In [ ]:
df.loc[df['day']=='Sat'].tip.mean()

2.993103448275862

In [ ]:
df.loc[df['day']=='Thur'].tip.mean()

2.771451612903225

In [ ]:
df.loc[df['day']=='Fri'].tip.mean()

2.7347368421052627

In [ ]:
# day ko aadhar ma tip ko mean nikalne
df.groupby(['day']).tip.mean()

day
Fri     2.734737
Sat     2.993103
Sun     3.255132
Thur    2.771452
Name: tip, dtype: float64

In [ ]:
df.groupby(['time']).tip.mean()

time
Dinner    3.102670
Lunch     2.728088
Name: tip, dtype: float64

In [ ]:
df.groupby(['smoker']).tip.mean()

smoker
No     2.991854
Yes    3.008710
Name: tip, dtype: float64

In [ ]:
df.groupby(['size']).tip.mean()

size
1    1.437500
2    2.582308
3    3.393158
4    4.135405
5    4.028000
6    5.225000
Name: tip, dtype: float64

In [ ]:
df.groupby(['sex','smoker']).tip.mean()

sex     smoker
Female  No        2.773519
        Yes       2.931515
Male    No        3.113402
        Yes       3.051167
Name: tip, dtype: float64

In [ ]:
# on average female smoker gives more tip
# on average male non smoker gives more tips

# loc vs iloc

In [ ]:
df.head(2)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3

In [ ]:
df.loc[12:14,['total_bill','tip']]

total_bill   tip
12       15.42  1.57
13       18.43  3.00
14       14.83  3.02

In [ ]:
# data slice garne and condition lagaune kam loc le garxa
# 14 row pani aayo ? loc vaneko location ho index hoina; loc le last ko value pani linxa


In [ ]:
# loc rule:
# 1. loc ma column ko thau ma index dina paudaina column name nai dinu parxa in lis
df.loc[12:14,0:2] # 0:2 lekhna paudaina loc expects column name instead not a index

TypeError: ignored

In [ ]:
# column ma name navayera index lekhnu paryo vane iloc use garne

In [ ]:
df.iloc[12:14, 0:2]

total_bill   tip
12       15.42  1.57
13       18.43  3.00

In [ ]:
# iloc le row ra column ma last ko value lidaina, iloc expects index in both row and column

In [ ]:
# last 2  column and first five rows
df.iloc[0:5,-2:]

time  size
0  Dinner     2
1  Dinner     3
2  Dinner     3
3  Dinner     2
4  Dinner     4